playwright/src/main/java/com/microsoft/playwright/impl/UrlMatcher.java

```diff
@@ -68,50 +64,75 @@ static String resolveUrl(URL baseUrl, String spec) {
     }
   }
 
-  UrlMatcher(URL base, String url) {
-    this(url, toPredicate(Pattern.compile(globToRegex(resolveUrl(base, url)))).or(s -> url == null || url.equals(s)));
+  UrlMatcher(URL baseURL, String glob) {
+    this(baseURL, glob, null, null);
   }
 
   UrlMatcher(Pattern pattern) {
-    this(pattern, toPredicate(pattern));
+    this(null, null, pattern, null);
   }
+
   UrlMatcher(Predicate<String> predicate) {
-    this(predicate, predicate);
+    this(null, null, null, predicate);
   }
 
-  private UrlMatcher(Object rawSource, Predicate<String> predicate) {
-    this.rawSource = rawSource;
+  private UrlMatcher(URL baseURL, String glob, Pattern pattern, Predicate<String> predicate) {
+    this.baseURL = baseURL;
+    this.glob = glob;
+    this.pattern = pattern;
     this.predicate = predicate;
   }
 
   boolean test(String value) {
-    return predicate == null || predicate.test(value);
+    if (pattern != null) {
+      return pattern.matcher(value).find();
+    }
+    if (predicate != null) {
+      return predicate.test(value);
+    }
+    if (glob != null) {
+      return Pattern.compile(globToRegex(resolveUrl(baseURL, glob))).matcher(value).find();
+    }
+    return true;
   }
 
   @Override
   public boolean equals(Object o) {
     if (this == o) return true;
     if (o == null || getClass() != o.getClass()) return false;
     UrlMatcher that = (UrlMatcher) o;
-    if (rawSource instanceof Pattern && that.rawSource instanceof Pattern) {
-      Pattern a = (Pattern) rawSource;
-      Pattern b = (Pattern) that.rawSource;
-      return a.pattern().equals(b.pattern()) && a.flags() == b.flags();
+    if (pattern != null) {
+      return that.pattern != null && pattern.pattern().equals(that.pattern.pattern()) && pattern.flags() == that.pattern.flags();
+    }
+    if (predicate != null) {
+      return predicate.equals(that.predicate);
+    }
+    if (glob != null) {
+      return glob.equals(that.glob);
     }
-    return Objects.equals(rawSource, that.rawSource);
+    return true;
   }
 
   @Override
   public int hashCode() {
-    return Objects.hash(rawSource);
+    if (pattern != null) {
+      return pattern.hashCode();
+    }
+    if (predicate != null) {
+      return predicate.hashCode();
+    }
+    if (glob != null) {
+      return glob.hashCode();
+    }
+    return super.hashCode();
```

Comment:
This will vary depending on the object instance and we want it to be a constant for `<any>` pattern, `return 1` or something?